Copyright (c) Microsoft Corporation.
Licensed under the MIT License.

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 98bbb85..349e1a4 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -172,6 +172,7 @@ set(CC_SOURCES
     src/misc/ibvwrap.cc
     src/misc/nvmlwrap_stub.cc
     src/misc/rocm_smi_wrap.cc
+    src/misc/npkit.cc
     src/transport/coll_net.cc
     src/transport/net.cc
     src/transport/net_ib.cc
@@ -204,6 +205,10 @@ if(TIMING_PROFILE)
   add_definitions(-DENABLE_TIMING_PROFILE)
 endif()
 
+if(NPKIT_FLAGS)
+  add_definitions(${NPKIT_FLAGS})
+endif()
+
 set(COLLTRACE 1 CACHE BOOL "Collective Trace Option")
 if(COLLTRACE)
   add_definitions(-DENABLE_COLLTRACE)
