Copyright (c) Microsoft Corporation.
Licensed under the MIT License.

diff --git a/src/channel.cc b/src/channel.cc
index b2cc5b7..fa2d2dd 100644
--- a/src/channel.cc
+++ b/src/channel.cc
@@ -6,6 +6,9 @@
 
 #include "channel.h"
 #include "param.h"
+#if defined(ENABLE_NPKIT)
+#include "npkit/npkit.h"
+#endif
 
 ncclResult_t initChannel(struct ncclComm* comm, int channelid) {
   struct ncclChannel* channel = comm->channels+channelid;
@@ -26,6 +29,15 @@ ncclResult_t initChannel(struct ncclComm* comm, int channelid) {
 
   // Per-channel operation list.
   NCCLCHECK(ncclCudaHostCalloc(&channel->workFifo, NCCL_MAX_OPS));
+
+#if defined(ENABLE_NPKIT)
+  // Init NpKit
+  channel->npKitEvent.bits[0] = channel->npKitEvent.bits[1] = 0;
+  channel->gpuNpKitEventCollectContext = NpKit::GetGpuEventCollectContext(channelid);
+  channel->cpuNpKitEventCollectContext = NpKit::GetCpuEventCollectContext(channelid);
+  channel->cpuTimestamp = NpKit::GetCpuTimeStamp();
+#endif
+
   return ncclSuccess;
 }
 
