Copyright (c) Microsoft Corporation.
Licensed under the MIT License.

diff --git a/src/init.cc b/src/init.cc
index 0367225..938765e 100644
--- a/src/init.cc
+++ b/src/init.cc
@@ -16,6 +16,9 @@
 #include "graph.h"
 #include "argcheck.h"
 #include "graph/topo.h"
+#if defined(ENABLE_NPKIT)
+#include "npkit/npkit.h"
+#endif
 #include <fcntl.h>
 #include <string.h>
 #include <errno.h>
@@ -288,6 +291,13 @@ static ncclResult_t devCommSetup(ncclComm_t comm) {
     comm->hostDevComm.scclAlgos[i] = comm->scclAlgos[i];
   comm->hostDevComm.scclAlgoShared = comm->scclAlgoShared;
   comm->hostDevComm.numberOfSCCAlgorithms = comm->numberOfSCCAlgorithms;
+
+#if defined(ENABLE_NPKIT)
+  // Init NPKit
+  NCCLCHECK(NpKit::Init(comm->rank));
+  comm->hostDevComm.npKitEventCollectContexts = NpKit::GetGpuEventCollectContexts();
+  comm->hostDevComm.cpuTimestamp = NpKit::GetCpuTimestamp();
+#endif
   
   // Duplicate the dev comm on the device
   NCCLCHECK(ncclCudaCalloc(&comm->devComm, 1));
@@ -1024,6 +1034,17 @@ static ncclResult_t commDestroy(ncclComm_t comm) {
 
   TRACE(NCCL_INIT, "Destroyed comm %p rank %d", comm, comm->rank);
 
+#if defined(ENABLE_NPKIT)
+  // Dump NPKit events and shutdown
+  const char* npkitDumpDir = getenv("NPKIT_DUMP_DIR");
+  if (npkitDumpDir == nullptr) {
+    WARN("NPKIT_DUMP_DIR is empty");
+  } else {
+    NCCLCHECK(NpKit::Dump(npkitDumpDir));
+  }
+  NCCLCHECK(NpKit::Shutdown());
+#endif
+
   return ncclSuccess;
 }
 
