Copyright (c) Microsoft Corporation.
Licensed under the MIT License.

diff --git a/src/include/proxy.h b/src/include/proxy.h
index 58a58b2..4053fe0 100644
--- a/src/include/proxy.h
+++ b/src/include/proxy.h
@@ -36,6 +36,10 @@ struct ncclProxySubArgs {
   uint64_t done;
   uint64_t end;
   void* requests[NCCL_STEPS];
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_NET_SEND_ENTRY) && defined(ENABLE_NPKIT_EVENT_NET_SEND_EXIT)
+  int npKitSizesFifo[NCCL_STEPS];
+#endif
 };
 
 struct ncclProxyArgs {
