Copyright (c) Microsoft Corporation.
Licensed under the MIT License.

diff --git a/src/collectives/device/all_reduce.h b/src/collectives/device/all_reduce.h
index 97f347a..8dc5fdd 100644
--- a/src/collectives/device/all_reduce.h
+++ b/src/collectives/device/all_reduce.h
@@ -9,6 +9,9 @@
 #include "collectives.h"
 #include "primitives.h"
 #include "clique/AllReduceCliqueKernel.h" // [RCCL] AllReduce Clique-based kernel support
+#if defined(ENABLE_NPKIT)
+#include "npkit/npkit.h"
+#endif
 
 namespace {
   template<typename T, typename RedOp, typename Proto>
@@ -29,6 +32,28 @@ namespace {
     if (tid == 0) clk = __builtin_amdgcn_s_memrealtime();
 #endif
 
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_TIME_SYNC_CPU)
+    if (threadIdx.x == 0) {
+      uint64_t* cpuTimestamp = ncclShmem->channel.cpuTimestamp;
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_TIME_SYNC_CPU, 0, 0, *cpuTimestamp,
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_TIME_SYNC_GPU)
+    if (threadIdx.x == 0) {
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_TIME_SYNC_GPU, 0, 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_ALL_REDUCE_RING_ENTRY)
+    if (threadIdx.x == 0) {
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_ALL_REDUCE_RING_ENTRY, size*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+
     int minChunkSize;
     if (Proto::Id == NCCL_PROTO_LL)
       minChunkSize = nthreads*(Proto::calcBytePerGrain()/sizeof(T));
@@ -112,6 +137,14 @@ namespace {
 #ifdef ENABLE_PROFILING
     if (tid == 0 && args->op.opCount) devProf->elems[blockIdx.x].total_cycle += (__builtin_amdgcn_s_memrealtime() - clk);
 #endif
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_ALL_REDUCE_RING_EXIT)
+    if (threadIdx.x == 0) {
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_ALL_REDUCE_RING_EXIT, size*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+
   }
 
   template<typename T, typename RedOp, typename Proto>
@@ -130,6 +163,28 @@ namespace {
     const ssize_t loopSize = int(nChannels*chunkSize);
     const ssize_t size = args->coll.count;
 
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_TIME_SYNC_CPU)
+    if (threadIdx.x == 0) {
+      uint64_t* cpuTimestamp = ncclShmem->channel.cpuTimestamp;
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_TIME_SYNC_CPU, 0, 0, *cpuTimestamp,
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_TIME_SYNC_GPU)
+    if (threadIdx.x == 0) {
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_TIME_SYNC_GPU, 0, 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_ALL_REDUCE_TREE_UPDOWN_ENTRY)
+    if (threadIdx.x == 0) {
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_ALL_REDUCE_TREE_UPDOWN_ENTRY, size*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+
     if (loopSize > size)
       chunkSize = divUp((int)size, int(nChannels*minChunkSize))*int(minChunkSize);
 
@@ -184,6 +239,14 @@ namespace {
         }
       }
     }
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_ALL_REDUCE_TREE_UPDOWN_EXIT)
+    if (threadIdx.x == 0) {
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_ALL_REDUCE_TREE_UPDOWN_EXIT, size*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+
   }
 
   template<typename T, typename RedOp, typename Proto>
@@ -213,6 +276,40 @@ namespace {
       nthreadsSplit = (nthreads*7/(10*WARP_SIZE))*WARP_SIZE;
     }
 
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_TIME_SYNC_CPU)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x == nthreadsSplit) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      uint64_t* cpuTimestamp = ncclShmem->channel.cpuTimestamp;
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_TIME_SYNC_CPU, 0, 0, *cpuTimestamp,
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_TIME_SYNC_GPU)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x == nthreadsSplit) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_TIME_SYNC_GPU, 0, 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_ALL_REDUCE_TREE_SPLIT_ENTRY)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x == nthreadsSplit) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_ALL_REDUCE_TREE_SPLIT_ENTRY, size*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+
     if (loopSize > size)
       chunkSize = divUp((int)size, nChannels*int(minChunkSize))*int(minChunkSize);
 
@@ -271,6 +368,18 @@ namespace {
         }
       }
     }
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_ALL_REDUCE_TREE_SPLIT_EXIT)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x == nthreadsSplit) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_ALL_REDUCE_TREE_SPLIT_EXIT, size*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+
   }
 }
 
