Copyright (c) Microsoft Corporation.
Licensed under the MIT License.

diff --git a/src/collectives/device/prims_ll128.h b/src/collectives/device/prims_ll128.h
index 81753a0..704e48a 100644
--- a/src/collectives/device/prims_ll128.h
+++ b/src/collectives/device/prims_ll128.h
@@ -6,6 +6,9 @@
  ************************************************************************/
 
 #include "op128.h"
+#if defined(ENABLE_NPKIT)
+#include "npkit/npkit.h"
+#endif
 
 #define NCCL_LL128_FLAGTHREAD (NCCL_LL128_LINEELEMS-1)
 
@@ -44,6 +47,13 @@ class Primitives<T, RedOp, Fan, Direct, ProtoLL128>:
   uint64_t* recvBuff[MaxRecv];
   uint64_t* sendBuff[MaxSend];
 
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_ENTRY) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_EXIT)
+  uint64_t npKitWaitRecvDataProcessSize = 0;
+  uint64_t npKitWaitRecvEntryTime = 0;
+  uint64_t npKitWaitRecvExitTime = 0;
+  uint64_t npKitWaitRecvTotalTime = 0;
+#endif
+
   inline __device__ int recvOffset(int i) { return (recvStep[i]%NCCL_STEPS)*stepSize; }
   inline __device__ int sendOffset(int i) { return (sendStep[i]%NCCL_STEPS)*stepSize; }
   inline __device__ uint64_t* recvPtr(int i) { return recvBuff[i]+recvOffset(i); }
@@ -67,6 +77,16 @@ class Primitives<T, RedOp, Fan, Direct, ProtoLL128>:
   }
 
   inline __device__ void waitSend(int nbytes) {
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_WAIT_SEND_ENTRY)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_PRIM_LL128_WAIT_SEND_ENTRY, nbytes, 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
     if (sendConnHeadPtr) {
       int spins = 0;
       while (sendConnHeadCache + NCCL_STEPS < sendConnHead + 1) {
@@ -78,6 +98,16 @@ class Primitives<T, RedOp, Fan, Direct, ProtoLL128>:
       }
       sendConnHead += 1;
     }
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_WAIT_SEND_EXIT)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_PRIM_LL128_WAIT_SEND_EXIT, nbytes, 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
   }
 
   inline __device__ void postRecv() {
@@ -189,6 +219,18 @@ class Primitives<T, RedOp, Fan, Direct, ProtoLL128>:
       uint64_t flag = recvFlag(0);
       bool needReload;
       int spins = 0;
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_ENTRY) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_EXIT)
+      int npkitWaitRecvSpins = 0;
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+      if (threadIdx.x != 0 && tid == 0) {
+#else
+      if (threadIdx.x == 0) {
+#endif
+        npKitWaitRecvEntryTime = __builtin_amdgcn_s_memrealtime();
+      }
+#endif
+
       do {
         needReload = false;
         #pragma unroll
@@ -196,7 +238,21 @@ class Primitives<T, RedOp, Fan, Direct, ProtoLL128>:
           load128(ptr+u*WARP_SIZE, vr[u], vr[u+1]);
           needReload |= flagThread && (vr[u+1] != flag);
         }
+        npkitWaitRecvSpins++;
       } while (__any_sync(WARP_MASK, needReload) && checkAbort(spins, 0, 0) == 0);
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_ENTRY) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_EXIT)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+      if (threadIdx.x != 0 && tid == 0) {
+#else
+      if (threadIdx.x == 0) {
+#endif
+        npKitWaitRecvExitTime = __builtin_amdgcn_s_memrealtime();
+        npKitWaitRecvTotalTime += (npKitWaitRecvExitTime - npKitWaitRecvEntryTime) * (npkitWaitRecvSpins - 1) / npkitWaitRecvSpins;
+        npkitWaitRecvSpins = 0;
+      }
+#endif
+
     }
 
     /************* Finish register load **************/
@@ -230,6 +286,18 @@ class Primitives<T, RedOp, Fan, Direct, ProtoLL128>:
         uint64_t* ptr = recvPtr(i)+ll128Offset;
         bool needReload;
         int spins = 0;
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_ENTRY) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_EXIT)
+        int npkitWaitRecvSpins = 0;
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+        if (threadIdx.x != 0 && tid == 0) {
+#else
+        if (threadIdx.x == 0) {
+#endif
+          npKitWaitRecvEntryTime = __builtin_amdgcn_s_memrealtime();
+        }
+#endif
+
         do {
           needReload = false;
           #pragma unroll
@@ -237,8 +305,21 @@ class Primitives<T, RedOp, Fan, Direct, ProtoLL128>:
             load128(ptr+u*WARP_SIZE, vr[u], vr[u+1]);
             needReload |= flagThread && (vr[u+1] != flag);
           }
+          npkitWaitRecvSpins++;
         } while (__any_sync(WARP_MASK, needReload) && checkAbort(spins, i, 0) == 0);
 
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_ENTRY) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_EXIT)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+        if (threadIdx.x != 0 && tid == 0) {
+#else
+        if (threadIdx.x == 0) {
+#endif
+          npKitWaitRecvExitTime = __builtin_amdgcn_s_memrealtime();
+          npKitWaitRecvTotalTime += (npKitWaitRecvExitTime - npKitWaitRecvEntryTime) * (npkitWaitRecvSpins - 1) / npkitWaitRecvSpins;
+          npkitWaitRecvSpins = 0;
+        }
+#endif
+
         #pragma unroll
         for (int u=0; u<ELEMS_PER_THREAD; u+=2) {
           v[u] = MULTI<RedOp, T>()(redOp, vr[u], v[u]);
@@ -299,6 +380,21 @@ class Primitives<T, RedOp, Fan, Direct, ProtoLL128>:
 
     if (SEND) waitSend(divUp(nelem, DataEltPerSlice)*WireWordPerSlice*sizeof(uint64_t));
     barrier();
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_ENTRY) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_EXIT)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      npKitWaitRecvTotalTime = 0;
+      npKitWaitRecvDataProcessSize = nelem*sizeof(T);
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_ENTRY,
+          npKitWaitRecvDataProcessSize, 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+
     nelem -= DataEltPerSlice*warp;
     srcPtr += DataEltPerSlice*warp;
     dstPtr += DataEltPerSlice*warp;
@@ -316,6 +412,19 @@ class Primitives<T, RedOp, Fan, Direct, ProtoLL128>:
     }
 
     barrier();
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_ENTRY) && defined(ENABLE_NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_EXIT)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_PRIM_LL128_DATA_PROCESS_EXIT,
+          npKitWaitRecvDataProcessSize, npKitWaitRecvTotalTime, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+
     if (SEND) for (int i=0; i < MaxSend; i++) sendStep[i] += 1;
     if (SEND) postSend();
     if (RECV) for (int i=0; i < MaxRecv; i++) recvStep[i] += 1;
@@ -401,27 +510,187 @@ public:
   }
 
   __device__ void send(intptr_t inpIx, int eltN) {
-    return GenericOp<0, 1, Input, -1>(inpIx, -1, eltN, false);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_SEND_ENTRY)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_SEND_ENTRY, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+    GenericOp<0, 1, Input, -1>(inpIx, -1, eltN, false);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_SEND_EXIT)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_SEND_EXIT, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
   }
   __device__ void sendFromOutput(intptr_t outIx, int eltN) {
-    return GenericOp<0, 1, Output, -1>(outIx, -1, eltN, false);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_SEND_FROM_OUTPUT_ENTRY)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_SEND_FROM_OUTPUT_ENTRY, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+    GenericOp<0, 1, Output, -1>(outIx, -1, eltN, false);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_SEND_FROM_OUTPUT_EXIT)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_SEND_FROM_OUTPUT_EXIT, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
   }
   __device__ void recv(intptr_t outIx, int eltN, bool postOp=false) {
-    return GenericOp<1, 0, -1, Output>(-1, outIx, eltN, postOp);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_RECV_ENTRY)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_RECV_ENTRY, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+    GenericOp<1, 0, -1, Output>(-1, outIx, eltN, postOp);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_RECV_EXIT)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_RECV_EXIT, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
   }
   __device__ void recvReduceSend(intptr_t inpIx, int eltN) {
-    return GenericOp<1, 1, Input, -1>(inpIx, -1, eltN, false);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_RECV_REDUCE_SEND_ENTRY)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_RECV_REDUCE_SEND_ENTRY, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+    GenericOp<1, 1, Input, -1>(inpIx, -1, eltN, false);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_RECV_REDUCE_SEND_EXIT)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_RECV_REDUCE_SEND_EXIT, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
   }
   __device__ void recvReduceCopy(intptr_t inpIx, intptr_t outIx, int eltN, bool postOp=false) {
-    return GenericOp<1, 0, Input, Output>(inpIx, outIx, eltN, postOp);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_RECV_REDUCE_COPY_ENTRY)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_RECV_REDUCE_COPY_ENTRY, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+    GenericOp<1, 0, Input, Output>(inpIx, outIx, eltN, postOp);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_RECV_REDUCE_COPY_EXIT)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_RECV_REDUCE_COPY_EXIT, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
   }
   __device__ void copySend(intptr_t inpIx, intptr_t outIx, int eltN, bool postOp=false) {
-    return GenericOp<0, 1, Input, Output>(inpIx, outIx, eltN, postOp);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_COPY_SEND_ENTRY)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_COPY_SEND_ENTRY, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+    GenericOp<0, 1, Input, Output>(inpIx, outIx, eltN, postOp);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_COPY_SEND_EXIT)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_COPY_SEND_EXIT, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
   }
   __device__ void recvCopySend(intptr_t outIx, int eltN, bool postOp=false) {
-    return GenericOp<1, 1, -1, Output>(-1, outIx, eltN, postOp);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_RECV_COPY_SEND_ENTRY)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_RECV_COPY_SEND_ENTRY, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+    GenericOp<1, 1, -1, Output>(-1, outIx, eltN, postOp);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_RECV_COPY_SEND_EXIT)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_RECV_COPY_SEND_EXIT, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
   }
   __device__ void recvReduceCopySend(intptr_t inpIx, intptr_t outIx, int eltN, bool postOp=false) {
-    return GenericOp<1, 1, Input, Output>(inpIx, outIx, eltN, postOp);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_RECV_REDUCE_COPY_SEND_ENTRY)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_RECV_REDUCE_COPY_SEND_ENTRY, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
+    GenericOp<1, 1, Input, Output>(inpIx, outIx, eltN, postOp);
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_RECV_REDUCE_COPY_SEND_EXIT)
+#if defined(ENABLE_NPKIT_THREAD_SPLIT_SECOND_HALF)
+    if (threadIdx.x != 0 && tid == 0) {
+#else
+    if (threadIdx.x == 0) {
+#endif
+      NpKit::GenerateAndCollectGpuEvent(NPKIT_EVENT_RECV_REDUCE_COPY_SEND_EXIT, eltN*sizeof(T), 0, __builtin_amdgcn_s_memrealtime(),
+          &(ncclShmem->channel.npKitEvent), ncclShmem->channel.gpuNpKitEventCollectContext);
+    }
+#endif
   }
 };
