Copyright (c) Microsoft Corporation.
Licensed under the MIT License.

diff --git a/src/include/proxy.h b/src/include/proxy.h
index 0f05154..b359dab 100644
--- a/src/include/proxy.h
+++ b/src/include/proxy.h
@@ -39,6 +39,10 @@ struct ncclProxyArgs {
   uint64_t done;
   uint64_t end;
   void* requests[NCCL_STEPS];
+
+#if defined(ENABLE_NPKIT) && defined(ENABLE_NPKIT_EVENT_NET_SEND_ENTRY) && defined(ENABLE_NPKIT_EVENT_NET_SEND_EXIT)
+  int npKitSizesFifo[NCCL_STEPS];
+#endif
   int idle;
 
   // Element linking
